### PR TITLE
Feature popup with refactored code

### PR DESCRIPTION
**Pull request by winson8222**
Separated Voice and Chat functions from game client. To setup just make the scene have the fields stated in the files.

For every scene, I suggest to have some standard methods (maybe create a new class)
1. setUpPlayer
2. setupEnemy
3. setupMap
4. setupCollision

Abstracted out client players movement sync
Fixed bug where Popup windows closes on any click

### CODE REVIEW

File: client/src/communications/SceneChat.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+// private ignoreNextClick: boolean = false;
+// private dialog: any;
+// TS check is disabled because the Game type scene is not defined in this file
+
+
+function SetUpSceneChat(scene: Phaser.Scene) {
+
+    scene.input.on('pointerdown', (pointer) => {
+        // Check if we should ignore scene click (the one that opens the dialog)
+        if (scene.ignoreNextClick) {
```
Comment (shashahchk) [inline client/src/communications/SceneChat.ts]: So when would we choose to purposely ignore scene click? Maybe add documentation for that

File: client/src/communications/SceneChat.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    let gameUIScene = scene.scene.get('game-ui');
+    scene.scene.get('game-ui').events.on('inputFocused', (isFocused) => {
+        console.log("keyboard disabled")
+        scene.input.keyboard.enabled = false;
+
+    });
+
+    scene.scene.get('game-ui').events.on('clickedOutside', (isFocused) => {
+        console.log("keyboard enabled")
+        scene.input.keyboard.enabled = true;
```
Comment (shashahchk) [inline client/src/communications/SceneChat.ts]: add some comments pls

File: client/src/communications/SceneChat.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    } else {
+        console.log("failed to set room")
+        // Handle the case where the scene might not be ready or needs to be launched
+    }
+}
+
+// This function checks if the user is typing in an input or textarea, put at the top of the update method
+// to stop other game logic from running if the user is typing
+// May need to be modified if you have other input types and textareas
+function CheckIfTyping() {
```
Comment (shashahchk) [inline client/src/communications/SceneChat.ts]: pls format the comment
This function does...
Needs to put at the top of UPDATE method...
May need to be modified
(gd to keep each point in separate lines thanks

File: client/src/communications/SceneCommunication.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+// private recorder: MediaRecorder | undefined;
+// private room: Colyseus.Room | undefined; //room is a property of the class
+// private xKey!: Phaser.Input.Keyboard.Key;
+// private mediaStream: MediaStream | undefined;
+// private recorderLimitTimeout = 0;
+// For reference to the scene
+// TS check is disabled because the Game type scene is not defined in this file
+
+
+
```
Comment (shashahchk) [inline client/src/communications/SceneCommunication.ts:28]: pretty minor but from. now on lets be careful with the empty lines and all styling

File: client/src/communications/SceneCommunication.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+function SetUpVoiceComm(scene: Phaser.Scene) {
+    getLocalStream()
+
+    scene.xKey.on('down', () => {
+        if (!scene.recorder && mediaStream) {
+            if (scene.room)
+                console.log("pushing")
```
Comment (shashahchk) [inline client/src/communications/SceneCommunication.ts]: lets make the logs more detailed, pushing recording.. or somethign like that, 
as we log more things it gets confusing

File: client/src/communications/SceneCommunication.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        });
+    };
+
+
+
+    console.log("start recording");
+    scene.recorder.start();
+
+    scene.recorderLimitTimeout = setTimeout(() => scene.recorder?.stop(), 10 * 1000);
+}//dt is the change since last frame
```
Comment (shashahchk) [inline client/src/communications/SceneCommunication.ts]: i think this comment is left by the previous update method. pls delete. 
Also same minor cosmetic issue, pls fix styling of entire file

File: client/src/scenes/DialogBox.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        }, this);
+    }
+
+    update() {}
+}
+
+var createDialog = function (scene, x, y, onClick) {
+
+
+
```
Comment (shashahchk) [inline client/src/scenes/DialogBox.ts]: pls fix styling

File: client/src/scenes/DialogBox.ts [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    var dialog = scene.rexUI.add.dialog({
+        x: x,
+        y: y,
+
+        background: scene.rexUI.add.roundRectangle(0, 0, 100, 100, 20, 0x0E376F),
+
+        title: scene.rexUI.add.label({
+            background: scene.rexUI.add.roundRectangle(0, 0, 100, 40, 20, 0x182456),
+            text: scene.add.text(0, 0, 'Difficulty: Simple', {
```
Comment (shashahchk) [inline client/src/scenes/DialogBox.ts]: try not to hardcode. define constants at top of file

File: client/src/scenes/DialogBox.ts [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        }
+    })
+        .layout()
+        .pushIntoBounds()
+        //.drawBounds(this.add.graphics(), 0xff0000)
+        .popUp(500);
+
+
+    dialog.on('button.click', function (button, groupName, index) {
+        if (button.name === 'fightButton') { // Check if the 'Fight' button was clicked
```
Comment (shashahchk) [inline client/src/scenes/DialogBox.ts]: try to not include comments that just explains the code (which can be inferred), try to explain the WHY more.

File: client/src/scenes/DialogBox.ts [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            console.log('Fight clicked');
+            // onclick call back
+        }
+    });
+
```
Comment (shashahchk) [inline client/src/scenes/DialogBox.ts]: styling issue. Please fix all files thank u!

File: client/src/scenes/Game.ts [hunk 11]
```diff
@@ -143,35 +182,59 @@ export default class Game extends Phaser.Scene {
...
 
+
+
+
+    }
+
+
+
+    // set up the map and the different layers to be added in the map for reference in collisionSetUp
+    private mapTileSetUp() {
```
Comment (shashahchk) [inline client/src/scenes/Game.ts]: minor but please name the functions that do something as verb
set_up_tilemap

File: client/src/scenes/Game.ts [hunk 12]
```diff
@@ -143,35 +182,59 @@ export default class Game extends Phaser.Scene {
...
+
+
         const map = this.make.tilemap({ key: 'user_room' })
```
Comment (shashahchk) [inline client/src/scenes/Game.ts]: why cant we use this map directly as a field?
this.map
not sure about whether it supports retrieval of layer

File: client/src/scenes/Game.ts [hunk 13]
```diff
@@ -143,35 +182,59 @@ export default class Game extends Phaser.Scene {
...
+    // create the enemies in the game, and design their behaviors
+    private createEnemies() {
+        return;
     }
 
-    update() {
+
+    update(t: number, dt: number)
+    {
+        //return if the user is typing
```
Comment (shashahchk) [inline client/src/scenes/Game.ts]: not sure what this ocmment means. This funciton si a bit messy

File: client/src/scenes/GameUi.ts [hunk 14]
```diff
@@ -1,17 +1,70 @@
...
+    private inputPanel: any;
+    private mainPanel: any;
+    private upperPanel: any;
+    private client: Colyseus.Client | undefined
+    private spaceKey: Phaser.Input.Keyboard.Key
+    private isFocused = false;
+    private inputBox: any;
+    private enterKey: Phaser.Input.Keyboard.Key;
+    private userNameBox: any;
+    // array of users
```
Comment (shashahchk) [inline client/src/scenes/GameUi.ts]: ?? dont see this

File: client/src/scenes/GameUi.ts [hunk 15]
```diff
@@ -1,17 +1,70 @@
...
-    create()
-    {
+    create() {
+
         const hearts = this.add.group({
             classType: Phaser.GameObjects.Image
         })
 
+        var userID = "Hello",
+            // i want to get the client's id
```
Comment (shashahchk) [inline client/src/scenes/GameUi.ts]: please remove comments like this

File: client/src/scenes/GameUi.ts [hunk 16]
```diff
@@ -20,6 +73,315 @@ export default class GameUi extends Phaser.Scene {
...
+            if (isOutside) {
+                // Emit an event or handle the outside click directly
+                this.events.emit('clickedOutside');
+                this.isFocused = false;
+            }
+        });
+
+
+        //temp fix
+        this.spaceKey.on('down', () => {
```
Comment (shashahchk) [inline client/src/scenes/GameUi.ts]: is there a way to make these comments more condensed and useful

File: client/src/scenes/GameUi.ts [hunk 17]
```diff
@@ -20,6 +73,315 @@ export default class GameUi extends Phaser.Scene {
...
+
+    setClient(client: Colyseus.Client) {
+        this.client = client;
+    }
+
+    messageToString(message) {
+        return `[${message.senderName}] ${message.message}\n`;
     }
+
+    setMessages(message) {
```
Comment (shashahchk) [inline client/src/scenes/GameUi.ts]: setMessage instead of setMessages

File: client/tsconfig.json [hunk 18]
```diff
@@ -2,7 +2,7 @@
 	"compilerOptions": {
 		"target": "es2016",
 		"module": "es6",
-		"strict": true,
+		"strict": false,
```
Comment (shashahchk) [inline client/tsconfig.json:5]: what does this do?

File: server/src/rooms/BattleRoom.ts [hunk 19]
```diff
@@ -1,21 +1,23 @@
...
     this.setState(new MyRoomState());
 
-    this.onMessage("keydown", (client, message) => {
-      //
-      this.broadcast('keydown', message, {
-        except: client
-      })
-      // handle "type" message
-      //
+    this.onMessage("push", (client, _) => {
```
Comment (shashahchk) [inline server/src/rooms/BattleRoom.ts]: what does push mean, can we be more specific in naming

File: client/src/scenes/Game.ts [hunk 20]
```diff
@@ -143,35 +182,59 @@ export default class Game extends Phaser.Scene {
...
-        //all animations are global once we add them
-        //set the body size of the sprite for collision handling
-        this.faune.body.setSize(this.faune.width * 0.5, this.faune.height * 0.8)
+    }
 
-        this.faune.anims.play('faune-idle-down')
 
-        this.cameras.main.startFollow(this.faune, true)
+    // set up the collision between different objects in the game
+    private collisionSetUp() {
```
Comment (shashahchk) [inline client/src/scenes/Game.ts]: please check all naming,
set_up_layers_collision

**Review by pygmychameleon — CHANGES_REQUESTED** [review]
Thanks for the hard work. It's quite a lot of code so I apologise if there are quite a lot of comments. Just want to make sure we standardise things early so problems don't arise later. The code is also quite complex so having insightful comments would help a lot so we can minimise having to bother you with questions down the line. 
Overall, looks solid, most issues are cosmetic.

File: client/tsconfig.json (hunk 18 above)
Comment (winson8222) [reply client/tsconfig.json:5]: This is to remove all the non type safe warning when calling methods on types. Example Phaser.scene may not have a certain method f but BattleRoom has, so need to suppress the warning that arise from scene.f() when the scene is used as an argument in another method.

File: client/src/scenes/DialogBox.ts (hunk 8 above)
Comment (winson8222) [reply client/src/scenes/DialogBox.ts]: just a sample will add more when have other variations

File: client/src/scenes/Game.ts (hunk 12 above)
Comment (winson8222) [reply client/src/scenes/Game.ts]: will look into this

**Comment by shashahchk** [comment]
LGTM!!!!!!!

**Pull request merged**